---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(DESeq2)
library(ggpubr)
library(readxl)
```

```{r}
counts <- read_tsv("../../bono_hirota_2020/salmon_human18.tsv")
counts
```

```{r}
hn2_chip <- read_tsv("../../bono_hirota_2020/human18_hn2_ChIP.tsv")
```
```{r}
rna_data <- read_tsv("../../bono_hirota_2020/rnaseq_human18.tsv")
for(i in 2:nrow(rna_data)){
  if(is.na(rna_data$human[i]))
    rna_data$human[i] <- rna_data$human[i-1]
}
rna_data$duration[is.na(rna_data$duration)] <- "unknown"
rna_data$`O2 concentration`[is.na(rna_data$`O2 concentration`)] <- "unknown"

rna_data_ps <- rna_data %>%
  select(human, sample, `O2 concentration`, duration, `hypoxia(SRR)`, `normoxia(SRR)`) %>%
  unite(1:4, col = "sample", sep = "_")

mult_rep <- rna_data %>%
  group_by(human, sample, `O2 concentration`, duration) %>%
  tally() %>%
  arrange(desc(n)) %>%
  filter(n > 1) %>%
  unite(1:4, col = "sample", sep = "_")

rna_data_ps %>%
  filter(sample %in% mult_rep$sample) %>%
  write_tsv("../../bono_hirota_2020/ps_filtered_data.tsv")
```

```{r}
deseq.res <- read_csv("/data2/2021_Rongo/data_frames/deseq_results.csv")
deseq.res <- deseq.res %>%
  separate(col = samples, into = c("data", "cell.type", "condition", "duration"), sep = "_", remove = F)

write_csv(deseq.res, gzfile("../output/tables/human_hypoxia_diff_expression.csv.gz"))s
```

## human orthologs
```{r}
tf.hits <- read_xlsx("../data/83 direct targets that have human homologs.xlsx")
tf.hits <- tf.hits %>% filter(`Human Ortholog` != "N.A.") # Remove genes with no human orthologs
tf.hits$ENSG <- sapply(strsplit(tf.hits$`Human Ortholog`, " "), `[[`, 1)
ortho <- tf.hits %>%
  select(ENSG) %>%
  distinct()

genes.deseq <- counts %>% 
  select(GeneName) %>% 
  dplyr::rename(Gene = GeneName)

gene.id <- read_tsv("../../bono_hirota_2020/170704all_HN2.txt") %>%
  select(ENSG_Gene) %>%
  separate(ENSG_Gene, into = c("ENSG", "Gene"), sep = "_", extra = "merge")

final.ortho <- left_join(genes.deseq, gene.id, by = "Gene") %>%
  left_join(ortho, ., by = "ENSG") %>%
  filter(!is.na(Gene)) %>%
  arrange(Gene) 

```

## Percentile of orthologs in gene expression changes
```{r}
ecdf_fun <- function(data, genes) {
  x <- data$log2FoldChange
  perc <- data$log2FoldChange[data$gene_name %in% genes]
  out <- ecdf(x)(perc)
  names(out) <- genes
  return(out)
  }

perc.ortho <- deseq.res %>%
  arrange(samples, gene_name) %>%
  group_by(samples, data, cell.type, condition, duration) %>%
  group_modify(~ ecdf_fun(data = ., genes = final.ortho$Gene) %>%
                 enframe(name = "Gene", value = "quantile"))

perc.ortho

write_csv(perc.ortho, "../output/tables/worm_orthologs_human_hypoxia_percentile.csv")
```

```{r fig.width=20, fig.height=15}
ortho.perc.1 <- ggplot(perc.ortho, aes(quantile)) +
  geom_density() +
  facet_wrap(~Gene, scales = "free_y") +
  theme_pubr()

ortho.perc.1 

ortho.perc.2 <- ggplot(perc.ortho, aes(quantile, fill = duration)) +
  geom_histogram() +
  scale_fill_brewer(palette = "Dark2") +
  facet_wrap(~Gene, scales = "free_y") +
  theme_pubr()

ortho.perc.2 

ortho.perc.3 <- ggplot(perc.ortho, aes(quantile, fill = condition)) +
  geom_histogram() +
  scale_fill_brewer(palette = "Dark2") +
  facet_wrap(~Gene, scales = "free_y") +
  theme_pubr()

ortho.perc.3 

ggsave(ortho.perc.1, filename = "../output/plots/plot.dist.percentile.human.hypoxia.ortholog.pdf", width = 20, height = 15)
ggsave(ortho.perc.2, filename = "../output/plots/plot.hist.percentile.duration.human.hypoxia.ortholog.pdf", width = 20, height = 15)
ggsave(ortho.perc.3, filename = "../output/plots/plot.hist.percentile.condition.human.hypoxia.ortholog.pdf", width = 20, height = 15)
```
```{r}
neigh.genes <- c("DHRS4L1",
"CARMIL3",
"CPNE6",
"NRL",
"DCAF11",
"FITM1",
"PSME1",
"EMC9",
"PSME2",
"RNF31")

perc.ortho.2 <- deseq.res %>%
  arrange(samples, gene_name) %>%
  group_by(samples, data, cell.type, condition, duration) %>%
  group_modify(~ ecdf_fun(data = ., genes = neigh.genes) %>%
                 enframe(name = "Gene", value = "quantile"))

perc.ortho
perc.ortho.2

write_csv(perc.ortho.2, "../output/tables/neighbors_worm_orthologs_human_hypoxia_percentile.csv")
```

```{r fig.width=20, fig.height=15}
neigh.ortho.perc.1 <- ggplot(perc.ortho.2, aes(quantile)) +
  geom_density() +
  facet_wrap(~Gene, scales = "free_y") +
  theme_pubr()

neigh.ortho.perc.1 

neigh.ortho.perc.2 <- ggplot(perc.ortho.2, aes(quantile, fill = duration)) +
  geom_histogram() +
  scale_fill_brewer(palette = "Dark2") +
  facet_wrap(~Gene, scales = "free_y") +
  theme_pubr()

neigh.ortho.perc.2 

neigh.ortho.perc.3 <- ggplot(perc.ortho.2, aes(quantile, fill = condition)) +
  geom_histogram() +
  scale_fill_brewer(palette = "Dark2") +
  facet_wrap(~Gene, scales = "free_y") +
  theme_pubr()

neigh.ortho.perc.3 

ggsave(neigh.ortho.perc.1, filename = "../output/plots/plot.dist.percentile.human.hypoxia.ortholog.neigh.pdf", width = 20, height = 15)
ggsave(neigh.ortho.perc.2, filename = "../output/plots/plot.hist.percentile.duration.human.hypoxia.ortholog.neigh.pdf", width = 20, height = 15)
ggsave(neigh.ortho.perc.3, filename = "../output/plots/plot.hist.percentile.condition.human.hypoxia.ortholog.neigh.pdf", width = 20, height = 15)
```




## Correlataion between orthologs
```{r fig.width = 10, fig.height = 10}
a <- perc.ortho %>%
  ungroup() %>%
  select(samples, Gene, quantile) %>%
  pivot_wider(names_from = Gene, values_from = quantile) %>%
  select(-samples) %>%
  cor(., use = "pairwise.complete.obs", method = "spearman")

a[upper.tri(a)] <- NA

b <- deseq.res %>%
  filter(gene_name %in% final.ortho$Gene) %>%
  select(samples, gene_name, log2FoldChange) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange) %>%
  select(-samples) %>%
  cor(., use = "pairwise.complete.obs", method = "spearman")

b[upper.tri(b)] <- NA

a2 <- a %>%
  as.data.frame() %>%
  rownames_to_column("From") %>%
  as_tibble() %>%
  pivot_longer(2:76, names_to = "To", values_to = "Cor") %>%
  filter(!is.na(Cor)) %>%
  filter(From != To, To!="TDRD15", From!="TDRD15")

a2
a2 %>% filter(Cor > 0.7, From != To, To!="TDRD15", From!="TDRD15") 
a2 %>% filter(Cor < -0.7, From != To, To!="TDRD15", From!="TDRD15") 

ggplot(a2, aes(From, To, fill = Cor)) +
  geom_tile() +
  scale_fill_gradient2(low ="red",
  mid = "white",
  high = "blue",
  midpoint = 0)

pheatmap::pheatmap(a)
pheatmap::pheatmap(b)
```

```{r}
dim(a)
```

```{r fig.width=20, fig.height=15}
perc.ortho %>%
  filter(quantile > 0.75) %>%
  group_by(Gene) %>%
  tally() %>%
  arrange(desc(n))
```




```{r fig.width=35, fig.height=10}
ortho.new <- read_xlsx("../../Orthologs of Worm HIF-1 ChIP-seq Direct Targets.xlsx")

d <- deseq.res %>%
  filter(gene_name %in% c(final.ortho$Gene, "PCK1")) %>%
  select(samples, gene_name, log2FoldChange) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange) %>%
  select(-samples)

pheatmap::pheatmap(d)

d <- deseq.res %>%
  filter(gene_name %in% ortho.new$`Gene Symbol`) %>%
  select(samples, gene_name, log2FoldChange) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange) 

x <- apply(d, 2, function(x)sum(is.na(x)))
d <- d[-24,-which(x>11)] 

my_gene_row <- d %>%
  select(samples) %>%
  separate(samples, into = c("GSE", "cell.type", "condition", "duration"), remove = F, sep = "_") %>%
  select(-GSE) %>% 
  column_to_rownames("samples") 
  

my_gene_row <- my_gene_row[-24,]

paletteLength <- 50
myColor <- colorRampPalette(c("blue", "white", "red"))(paletteLength)
# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd length pallettelengths
myBreaks <- c(seq(-10, 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(10/paletteLength, 10, length.out=floor(paletteLength/2)))


# category.colors <- colorRampPalette(grDevices::rainbow(30))
# cell.type.col <- category.colors(length(unique(my_gene_row$cell.type)))
# names(cell.type.col) <- unique(my_gene_row$cell.type)
# condition.col <- category.colors(length(unique(my_gene_row$condition)))
# names(condition.col) <- unique(my_gene_row$condition)
# duration.col <- category.colors(length(unique(my_gene_row$duration)))
# names(duration.col) <- unique(my_gene_row$duration)

# mycolors <- list(cell.type = cell.type.col, condition = condition.col, duration = duration.col)

d <- d %>% 
  column_to_rownames("samples") 
# Plot the heatmap
# ortho.heatmap <- pheatmap::pheatmap(d, color = myColor, breaks = myBreaks, annotation_row = my_gene_row, annotation_colors = mycolors)
callback = function(hc, ...){dendsort(hc)}
ortho.heatmap <- pheatmap::pheatmap(d, color = myColor, breaks = myBreaks, clustering_callback = callback)

ggsave(ortho.heatmap, filename = "../output/plots/plot.heatmap.human.hypoxia.orthologs.pdf", width = 35, height = 10)

# pheatmap::pheatmap(d)

```

```{r fig.width=35, fig.height=10}
callback = function(hc, ...){dendsort(hc)}
pheatmap::pheatmap(d, color = myColor, breaks = myBreaks, annotation_row = my_gene_row, annotation_colors = mycolors, clustering_callback = callback)
```


```{r}
deseq.res %>%
  filter(gene_name %in% ortho.new$`Gene Symbol`) %>%
  select(samples, gene_name, log2FoldChange) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange) %>%
  .[24,]
```


```{r}
d <- deseq.res %>%
  filter(gene_name %in% ortho.new$`Gene Symbol`) %>%
  select(samples, gene_name, log2FoldChange) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange)

d %>%
  select(samples, PCK1, PCK2) %>%
  arrange(desc(PCK2)) %>%
  filter(grepl("CoCl2", samples))
```

```{r fig.width=25, fig.height=10}
q <- deseq.res %>%
  # filter(gene_name %in% ortho.new$`Gene Symbol`) %>%
  select(samples, gene_name, log2FoldChange) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange) %>%
  select(-samples)

x <- apply(q, 2, function(x)sum(is.na(x)))
q <- q[,-which(x>11)]


paletteLength <- 50
myColor <- colorRampPalette(c("blue", "white", "red"))(paletteLength)
# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd length pallettelengths
myBreaks <- c(seq(-20, 0, length.out=ceiling(paletteLength/2) + 1), 
              seq(20/paletteLength, 10, length.out=floor(paletteLength/2)))

# Plot the heatmap
pheatmap::pheatmap(q, color=myColor, breaks=myBreaks)
```


```{r}
deseq.res %>%
  filter(log2FoldChange > 0, grepl("PCK2", gene_name)) %>%
  arrange(desc(log2FoldChange))
```

```{r}
a <- read_tsv("../../HIF1A.10.tsv")
b <- a %>%
  filter(grepl("PCK", Target_genes)) %>%
  select(1,2)
unlist(b[,2])
which(b>0)

ggplot(a, aes(`HIF1A|Average`)) +
  geom_density() +
  scale_x_log10() +
  geom_vline(xintercept = unlist(b[,2]), inherit.aes = F)

ecdf(unlist(a[,2]))(unlist(b[,2]))
```





```{r}
obs.counts.hist <- deseq.res %>%
  filter(gene_name %in% ortho.new$`Gene Symbol`) %>%
  select(samples, gene_name, log2FoldChange) %>%
  group_by(gene_name) %>% 
  mutate(m = sum(!is.na(log2FoldChange))) %>%
  select(gene_name, m) %>%
  distinct() %>%
  ggplot(., aes(m)) + 
  geom_histogram(binwidth = 1) +
  theme_pubr() +
  labs(x = "Number of datasets with obsserved expression", y = "number of genes")
obs.counts.hist

ggsave(obs.counts.hist, filename = "../output/plots/plot.dist.human.hypoxia.orthologs.obs.datasets.pdf")
```

```{r fig.width = 30, fig.height=30}
cor.ortho <- deseq.res %>%
  filter(gene_name %in% ortho.new$`Gene Symbol`) %>%
  select(samples, gene_name, log2FoldChange) %>%
  group_by(gene_name) %>% 
  mutate(m = sum(!is.na(log2FoldChange))) %>%
  ungroup() %>%
  # filter(m >= 15) %>%
  select(-m) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange) %>%
  select(-samples) %>%
  cor(., use = "pairwise.complete.obs", method = "spearman")

paletteLength <- 50
myColor <- colorRampPalette(c("#0571b0", "white", "#ca0020"))(paletteLength)
myBreaks <- c(seq(-1, 1, length.out= paletteLength-1))

# Plot the heatmap
all.cor.fc <- pheatmap::pheatmap(cor.ortho, color=myColor, breaks=myBreaks)
all.cor.fc

cor.ortho <- deseq.res %>%
  filter(gene_name %in% ortho.new$`Gene Symbol`) %>%
  select(samples, gene_name, log2FoldChange) %>%
  group_by(gene_name) %>% 
  mutate(m = sum(!is.na(log2FoldChange))) %>%
  ungroup() %>%
  filter(m >= 15) %>%
  select(-m) %>%
  pivot_wider(names_from = gene_name, values_from = log2FoldChange) %>%
  select(-samples) %>%
  cor(., use = "pairwise.complete.obs", method = "spearman")

# cor.ortho[upper.tri(cor.ortho)] <- NA

only15.cor.fc <- pheatmap::pheatmap(cor.ortho, color=myColor, breaks=myBreaks)
only15.cor.fc

ggsave(all.cor.fc, filename = "../output/plots/plot.heatmap.cor.human.hypoxia.orthologs.all.genes.pdf", width = 30, height = 30)
ggsave(only15.cor.fc, filename = "../output/plots/plot.heatmap.cor.human.hypoxia.orthologs.gt15data.genes.pdf", width = 30, height = 30)

```


```{r}
all.cor.fc
```


```{r}
a <- deseq.res %>%
  group_by(gene_name) %>%
  summarise(pos1.l2fc = sum(log2FoldChange > 1, na.rm = T),
            pos0.l2fc = sum(log2FoldChange > 0, na.rm = T),
            possig.l2fc = sum(log2FoldChange > 0 & padj < 0.05, na.rm = T),
            med.l2fc = median(log2FoldChange, na.rm = T)) %>%
  left_join(., ortho.new %>% dplyr::rename(gene_name = `Gene Symbol`) %>% mutate(ortho = TRUE), by = "gene_name")

a
```

```{r}
a %>% arrange(desc(possig.l2fc), desc(pos1.l2fc), desc(pos0.l2fc))

a %>%
  ggplot(., aes(pos.l2fc)) +
  geom_bar() +
  xlim(10, 33)


```


```{r}
library("ggVennDiagram")
```

```{r}
x <- list(all = a %>% filter(possig.l2fc >= 20) %>% pull(gene_name),
          ortho = a %>% filter(ortho == T) %>% pull(gene_name))
ggVennDiagram(x)
```

```{r}
x <- list(all.up = a %>% filter(pos0.l2fc >= 15) %>% pull(gene_name),
          all.sig.up = a %>% filter(possig.l2fc >= 15) %>% pull(gene_name),
          ortho.up = a %>% filter(ortho == T, pos0.l2fc >= 15) %>% pull(gene_name),
          ortho.sig.up = a %>% filter(ortho == T, possig.l2fc >= 15) %>% pull(gene_name))

display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

display_venn(
        x,
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = c("#999999", "#E69F00", "#56B4E9", "#009E73"),
        # Numbers
        cex = 1,
        # Set names
        cat.cex = 1.2,
        cat.default.pos = "outer",
        cat.fontface = "bold"
)
# dev.off()
```

```{r}
a

nrow(a)
sum(a$pos0.l2fc >= 15)
sum(a$pos0.l2fc >= 15 & !is.na(a$ortho))
sum(!is.na(a$ortho))

sum(a$possig.l2fc >= 15)
sum(a$possig.l2fc >= 15 & !is.na(a$ortho))

```

```{r}
prop.test(x = 109, n = 235, p = 11156/37268)
prop.test(x = 16, n = 235, p = 135/37268)
```


```{r}
a %>% filter(a$possig.l2fc >= 15 & is.na(a$ortho)) %>% select(gene_name) %>% write_csv("bono_signif_upreg_15data_not-ortho.csv")
```




