---
title: "Overlap between HIF-1 ChIP peaks and modENCODE data"
output: html_notebook
---

### Packages
```{r}
library(readxl)
library(tidyverse)
library(GenomicRanges)
library(ggpubr)
```

```{r}
## Read in HIF1 binding sites
hif1 <- read_xlsx("../data/Peak coordinates for 3 conditions.xlsx", sheet = "OR3350 Peaks")

colnames(hif1) <- c("chr","start","end")
hif1 <- hif1 %>% 
  distinct() %>% # Remove duplicate peaks
  arrange(chr, start, end)
hif1$chr[hif1$chr=="MtDNA"] <- "chrMtDNA"

# Convert the hif-1 peaks into a GRanges object
hif1.gr.ini <- GRanges(seqnames = hif1$chr, ranges = IRanges(start = hif1$start, end = hif1$end), strand = "+")

# Combine overlapping peaks (all overlaps are >75%)
hif1.gr <- GenomicRanges::reduce(hif1.gr.ini, with.revmap = T)
```


```{r}
## Read in modENCODE all TF binding sitess
modTF <- read_xlsx("../data/TF binding coordinates for ALL worm TF by ModEncode at L4 Stage- Kudron et all 2018 .xlsx")
modTF <- modTF %>% filter(`#TF` != "HIF-1") # Peaks of other TF in modENCODE data

# Convert the modENCODE peaks into a GRanges object
modTF.gr <- GRanges(seqnames = modTF$`#Chromosome`, ranges = IRanges(start = modTF$`#peak_Start`, end = modTF$`#Peak_end`), strand = "+")
mcols(modTF.gr) <- list(TF = modTF$`#TF`)
```

### Function for processing overlaps
```{r}
overlap_processor <- function(data, prefix, hif){ 
  data %>%
    mutate(TF = modTF$`#TF`[subjectHits]) %>% # Identity of TF hits in modENCODE data
    group_by(queryHits, TF) %>%  
    tally() %>% # Count number of times a HIF-1 site is overlapping with each of the other TF
    ungroup() %>% 
    dplyr::rename(id = queryHits) %>% # Rename columns
    pivot_wider(names_from = TF, values_from = n, names_prefix = prefix) %>% # Rearrange data
    left_join(hif %>% 
                as_tibble() %>% 
                select(-strand) %>%
                mutate(id = 1:nrow(.)), ., by = "id") %>% # Join with HIF-1 positions
    rowwise() %>% 
    mutate(Total = sum(c_across(contains(prefix)), na.rm=T)) %>% # Calculate total number of hits per HIF-1 site
    select(seqnames, start, end, Total, everything(), -c(id, width)) %>% # Rearrange columns and remove a few unneeded ones
    dplyr::rename(chr = seqnames) %>%
    ungroup()
}
```

### Find overlaps between HIF-sites and HOT sites
```{r}
## Find if midpoint of hif1 binding site is
### a) within 400 bp of HOT sites
### b) within 1kb of HOT sites

modTF.400.gr <- resize(modTF.gr, width = 400, fix = "center")
modTF.1k.gr <- resize(modTF.gr, width = 1000, fix = "center")

#-------------------------------------------------------------- 
## Experimental data with other TFs in modENCODE data
hif1.mid.gr <- hif1.gr
start(hif1.mid.gr) <- floor(start(hif1.gr) + (end(hif1.gr)-start(hif1.gr))/2)
end(hif1.mid.gr) <- floor(start(hif1.gr) + (end(hif1.gr)-start(hif1.gr))/2)

## Find and process overlaps
over400 <- findOverlaps(hif1.mid.gr, modTF.400.gr) %>%
  as_tibble() %>%
  overlap_processor(prefix = "400-", hif = hif1.gr)
over1k <- findOverlaps(hif1.mid.gr, modTF.1k.gr) %>%
  as_tibble() %>%
  overlap_processor(prefix = "1k-", hif = hif1.gr)

```

```{r}
### Prepare files for Mehul/Chris

outExp <- left_join(over400, over1k, by = c("chr", "start", "end"), 
                    suffix = c(".400", ".1k")) %>%
  select(chr, start, end, Total.400, Total.1k, everything())

write.csv(outExp, "../output/tables/HIF-1_ModEncode_overlap.csv")
```


```{r fig.width=15,fig.height = 5}
beta.peaks <- read_xlsx("../data/Peak coordinates for 3 conditions.xlsx", sheet = "BETA Peaks")

beta.gr <- GRanges(seqnames = beta.peaks$chrom, ranges = IRanges(start = beta.peaks$pStart, end = beta.peaks$pEnd), strand = "+")
tmp.gr <- GRanges(seqnames = outExp$chr, ranges = IRanges(start = outExp$start, end = outExp$end), strand = "+")

# Find which peaks in the HIF-1 binding data are BETA peaks
overlaps.with.mod.data <- findOverlaps(beta.gr, tmp.gr)

# Get other TF binding data for BETA peaks
beta.data <- outExp[subjectHits(overlaps.with.mod.data), ]

# Distribution of number of times a TF is associated with a HIF-1 peak
TF.sum.perHIF <- beta.data %>%
  select(starts_with("400")) %>% # Consider TF binding within 400bp of BETA peaks
  pivot_longer(cols = starts_with("400"), names_to = "TF", values_to = "CountPerHIF") %>%
  filter(!is.na(CountPerHIF)) %>%
  group_by(TF) %>%
  summarise(total = sum(CountPerHIF)) %>%
  arrange(total) %>%
  mutate(TF = substr(TF, 5, nchar(TF)),
         TF = factor(TF, levels = TF))


plot.TF.sum.perHIF <- ggplot(TF.sum.perHIF, aes(x = TF, y = total)) +
  geom_bar(stat = "identity") +
  labs(x = "Transcription Factor", y = "Total instances of TF bound at a HIF-1 ChIP-seq peak") +
  theme_pubr(x.text.angle = 90)
plot.TF.sum.perHIF
ggsave(plot.TF.sum.perHIF, filename = "../output/plots/plot.TF.sum.perHIF.BETA.pdf")

```

## Enrichment of other TF association with HIF-1 BETA peaks
```{r fig.width=15,fig.height = 5}
# Frequency of different TF-binding in modENCODE data
all.TF.counts <- modTF %>% 
  group_by(`#TF`) %>% 
  tally() %>% 
  arrange(desc(n)) %>%
  dplyr::rename(bg = n,
         TF = `#TF`)

TF.enrich.perHIF <- left_join(TF.sum.perHIF, all.TF.counts, by = "TF")

# Calculate enrichment of other TF association with HIF-1 BETA peaks
TF.enrich.perHIF <- TF.enrich.perHIF %>%
  mutate(enrich = (total/sum(total)/(bg/sum(bg)))) %>%
  arrange(enrich) %>%
  mutate(TF = factor(TF, levels = TF))
  
plot.TF.enrich.perHIF <- ggplot(TF.enrich.perHIF, aes(x = TF, y = enrich)) +
  geom_bar(stat = "identity") +
  labs(x = "Transcription Factor", y = "TF enrichment at a HIF-1 ChIP-seq peak") +
  scale_y_continuous(trans = "log2") +
 # theme()
  theme_pubr(x.text.angle = 90)
plot.TF.enrich.perHIF
ggsave(plot.TF.enrich.perHIF, filename = "../output/plots/plot.TF.enrich.perHIF.BETA.pdf")
```

## Enrichment of other TF association with HIF-1 BETA peaks with bootstrap intervals
```{r}
obs <- sum(TF.enrich.perHIF$total) # Total number of other TF binding sites associated with HIF-1
wgt <- TF.enrich.perHIF$bg # Background frequency of TF binding sites in modENCODE data
tf <- TF.enrich.perHIF$TF # Other TF names

sample.TF.freq <- function(x, obs, wgt){
  s <- sample(tf, size = obs, replace = TRUE, prob = wgt)
  tab <- table(factor(s, levels = tf))
  (tab/sum(tab))/(wgt/sum(wgt))
}
bg.enrich <- sapply(1:100000, sample.TF.freq, obs = obs, wgt = wgt) # Enrichment expected based on sampling of modENCODE TF frequency

boot <- t(apply(bg.enrich, 1, quantile, probs = c(0.005,0.995))) %>%  # Calculate 99% confidence intervals for expected enrichment
  as.data.frame() %>% 
  rownames_to_column(var = "TF") %>% 
  as_tibble() %>%
  rename(low = `0.5%`, high = `99.5%`)

boot$low[boot$low == 0] <- min(boot$low[boot$low > 0]) # Set lower-bound > 0 for visualization
```

```{r}
# Calculate enrichment of other TF association with HIF-1 BETA peaks
TF.enrich.perHIF <- left_join(TF.sum.perHIF, all.TF.counts, by = "TF")

TF.enrich.perHIF <- TF.enrich.perHIF %>%
  mutate(enrich = (total/sum(total)/(bg/sum(bg)))) %>%
  arrange(enrich) %>%
  mutate(TF = factor(TF, levels = TF))

TF.enrich.perHIF <- left_join(TF.enrich.perHIF, boot) %>%
  mutate(TF = factor(TF, levels = TF),
         'Significantly enriched' = ifelse(enrich < low | enrich > high, "TRUE", "FALSE"))

write.csv(TF.enrich.perHIF, "../output/tables/HIF-1_ModEncode_overlap.enrich.csv")
```

```{r fig.width=15,fig.height = 5}
plot.TF.enrich.perHIF.bti <- ggplot(TF.enrich.perHIF %>% 
                                      arrange(enrich) %>%
                                      mutate(TF = factor(TF, levels = TF)), aes(x = TF, y = enrich, fill = `Significantly enriched`)) +
  geom_bar(stat = "identity") +
  geom_errorbar(aes(ymin=low, ymax=high), alpha = 0.5) +
  scale_fill_brewer(palette = "Set1", direction = -1) +
  labs(x = "Transcription Factor", y = "TF enrichment at a HIF-1 ChIP-seq peak") +
  scale_y_continuous(trans = "log2", 
                     labels = scales::trans_format("log2", scales::math_format(2^.x))) +
  theme_pubr(x.text.angle = 90)
plot.TF.enrich.perHIF.bti
ggsave(plot.TF.enrich.perHIF.bti, filename = "../output/plots/plot.TF.enrich.with.bti.perHIF.BETA.pdf")

plot.TF.enrich.perHIF.bti.ord <- ggplot(TF.enrich.perHIF %>% 
                                      arrange(total) %>%
                                      mutate(TF = factor(TF, levels = TF)), aes(x = TF, y = enrich, fill = `Significantly enriched`)) +
  geom_bar(stat = "identity") +
  geom_errorbar(aes(ymin=low, ymax=high), alpha = 0.5) +
  scale_fill_brewer(palette = "Set1", direction = -1) +
  labs(x = "Transcription Factor", y = "TF enrichment at a HIF-1 ChIP-seq peak") +
  scale_y_continuous(trans = "log2", 
                     labels = scales::trans_format("log2", scales::math_format(2^.x))) +
  theme_pubr(x.text.angle = 90)
plot.TF.enrich.perHIF.bti.ord
ggsave(plot.TF.enrich.perHIF.bti.ord, filename = "../output/plots/plot.TF.enrich.with.bti.perHIF.BETA.ordered.pdf")
```

```{r fig.width=5,fig.height = 5}
plot.TF.enrich.vs.TF.freq <- ggplot(TF.enrich.perHIF, aes(x = total, y = enrich, color = `Significantly enriched`)) +
  geom_point() +
  scale_color_brewer(palette = "Set1", direction = -1) +
  labs(x = "TF frequency in modENCODE", y = "TF enrichment at a HIF-1 ChIP-seq peak") +
  scale_y_continuous(trans = "log2", 
                     labels = scales::trans_format("log2", scales::math_format(2^.x))) +
  theme_pubr()
plot.TF.enrich.vs.TF.freq
ggsave(plot.TF.enrich.vs.TF.freq, filename = "../output/plots/plot.TF.enrich.vs.TF.freq.pdf")
```

